File: Tests/ConnectorControlCoreTests/CollectionTests.swift
import XCTest
@testable import ConnectorControlCore

/// Mirror: windows/tests/ConnectorControl.Core.Tests/CollectionTests.cs
final class CollectionTests: XCTestCase {
    private func entry(_ url: String) -> MCPEntry {
        MCPEntry(config: RemotePattern.make(url: url))
    }

    // MARK: - Decoding

    func testV2RoundTripPreservesTwoCollections() throws {
        let store = MasterStore(
            activeCollection: "Work",
            collections: [
                "Work": Collection(mcps: ["a": entry("https://a.example/mcp")]),
                "Personal": Collection(mcps: ["b": entry("https://b.example/mcp")]),
            ])
        let data = try JSONEncoder().encode(store)
        let decoded = try JSONDecoder().decode(MasterStore.self, from: data)
        XCTAssertEqual(decoded, store)
        XCTAssertEqual(decoded.activeCollection, "Work")
        XCTAssertEqual(decoded.collections["Work"]?.mcps.keys.sorted(), ["a"])
        XCTAssertEqual(decoded.collections["Personal"]?.mcps.keys.sorted(), ["b"])
    }

    // MARK: - mcps accessor scoping

    func testMcpsAccessorReadsAndWritesOnlyActiveCollection() {
        var store = MasterStore(
            activeCollection: "Work",
            collections: [
                "Work": Collection(mcps: ["a": entry("https://a.example/mcp")]),
                "Personal": Collection(mcps: ["b": entry("https://b.example/mcp")]),
            ])
        store.mcps["c"] = entry("https://c.example/mcp")
        XCTAssertEqual(store.collections["Work"]?.mcps.keys.sorted(), ["a", "c"])
        XCTAssertEqual(store.collections["Personal"]?.mcps.keys.sorted(), ["b"], "untouched")
    }

    // MARK: - Collection management

    func testAddCollectionCopyingCurrent() {
        var store = MasterStore.single(["a": entry("https://a.example/mcp")])
        let error = store.addCollection(named: "Copy", copyingCurrent: true)
        XCTAssertNil(error)
        XCTAssertEqual(store.activeCollection, "Copy")
        XCTAssertEqual(store.collections["Copy"]?.mcps.keys.sorted(), ["a"])
        XCTAssertEqual(store.collections["Default"]?.mcps.keys.sorted(), ["a"], "original untouched")
    }

    func testAddCollectionEmptyStartsBlank() {
        var store = MasterStore.single(["a": entry("https://a.example/mcp")])
        let error = store.addCollection(named: "Fresh", copyingCurrent: false)
        XCTAssertNil(error)
        XCTAssertEqual(store.collections["Fresh"]?.mcps, [:])
    }

    func testAddCollectionWithoutActivatingLeavesTheActiveOne() {
        var store = MasterStore.single(["a": entry("https://a.example/mcp")])
        XCTAssertNil(store.addCollection(named: "Fresh", copyingCurrent: false, activating: false))
        XCTAssertEqual(store.collections["Fresh"]?.mcps, [:])
        XCTAssertEqual(store.activeCollection, "Default")
    }

    func testAddCollectionRejectsEmptyName() {
        var store = MasterStore.empty
        XCTAssertEqual(store.addCollection(named: "   ", copyingCurrent: false), "Name must not be empty.")
    }

    func testAddCollectionRejectsDuplicateName() {
        var store = MasterStore.empty
        XCTAssertEqual(store.addCollection(named: "Default", copyingCurrent: false),
                       "A collection named “Default” already exists.")
    }

    func testRenameActiveCollection() {
        var store = MasterStore.empty
        let error = store.renameCollection(store.activeCollection, to: "Main")
        XCTAssertNil(error)
        XCTAssertEqual(store.activeCollection, "Main")
        XCTAssertEqual(Array(store.collections.keys), ["Main"])
    }

    func testRenameActiveCollectionRejectsCollision() {
        var store = MasterStore(
            activeCollection: "Work",
            collections: ["Work": Collection(), "Personal": Collection()])
        XCTAssertEqual(store.renameCollection(store.activeCollection, to: "Personal"), "A collection named “Personal” already exists.")
        XCTAssertEqual(store.activeCollection, "Work", "unchanged on error")
    }

    func testRenameActiveCollectionRejectsEmptyName() {
        var store = MasterStore.empty
        XCTAssertNotNil(store.renameCollection(store.activeCollection, to: "  "))
    }

    func testDeleteActiveCollectionSwitchesToFirstRemaining() {
        var store = MasterStore(
            activeCollection: "Work",
            collections: ["Work": Collection(), "Alpha": Collection(), "Zeta": Collection()])
        let error = store.deleteCollection(named: store.activeCollection)
        XCTAssertNil(error)
        XCTAssertEqual(store.activeCollection, "Alpha")
        XCTAssertNil(store.collections["Work"])
    }

    func testDeleteActiveCollectionRejectsLastCollection() {
        var store = MasterStore.empty
        XCTAssertEqual(store.deleteCollection(named: store.activeCollection), "Can’t delete the last collection.")
        XCTAssertEqual(store.collections.count, 1)
    }

    func testSwitchCollection() {
        var store = MasterStore(
            activeCollection: "Work",
            collections: ["Work": Collection(), "Personal": Collection()])
        XCTAssertNil(store.switchCollection(to: "Personal"))
        XCTAssertEqual(store.activeCollection, "Personal")
    }

    func testSwitchCollectionRejectsUnknownName() {
        var store = MasterStore.empty
        XCTAssertEqual(store.switchCollection(to: "Nope"), "No collection named “Nope”.")
        XCTAssertEqual(store.activeCollection, "Default")
    }

    func testErrorMessagesUseTypographicPunctuationLikeTheMacApp() {
        var store = MasterStore.empty
        let duplicate = store.addCollection(named: "Default", copyingCurrent: false)!
        XCTAssertEqual(duplicate[duplicate.index(before: duplicate.range(of: "Default")!.lowerBound)], "“")
        XCTAssertEqual(duplicate[duplicate.range(of: "Default")!.upperBound], "”")
        XCTAssertTrue(store.deleteCollection(named: store.activeCollection)!.contains("’"))
    }
}

File: windows/tests/ConnectorControl.Core.Tests/CollectionTests.cs
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests;

public class CollectionTests : IDisposable
{
    private readonly TempDir dir = new("collection-store");
    private string Url => dir.File("mcps.json");

    public void Dispose() => dir.Dispose();

    private static McpEntry Entry(string url) => new(JsonValue.Object(
        ("command", JsonValue.String("npx")),
        ("args", JsonValue.Array([JsonValue.String("-y"), JsonValue.String("mcp-remote"), JsonValue.String(url)]))));

    private static string[] Keys(Collection? c) => c is null ? [] : c.Mcps.Keys.Order(StringComparer.Ordinal).ToArray();

    private static MasterStore TwoCollections() => new(2, "Work",
    [
        new("Work", new Collection(new Dictionary<string, McpEntry> { ["a"] = Entry("https://a.example/mcp") })),
        new("Personal", new Collection(new Dictionary<string, McpEntry> { ["b"] = Entry("https://b.example/mcp") })),
    ]);

    // Decoding

    [Fact]
    public void V2RoundTripPreservesTwoCollections()
    {
        var store = TwoCollections();
        var decoded = MasterStore.FromJson(JsonValue.Parse(store.ToJson().Serialize()));
        Assert.Equal(store, decoded);
        Assert.Equal("Work", decoded.ActiveCollection);
        Assert.Equal(["a"], Keys(decoded.Collections["Work"]));
        Assert.Equal(["b"], Keys(decoded.Collections["Personal"]));
    }

    [Fact]
    public void UnknownActiveCollectionFallsBackToExistingCollection()
    {
        File.WriteAllText(Url, "{\"version\":2,\"activeProfile\":\"Ghost\",\"profiles\":{\"Alpha\":{\"mcps\":{}},\"Beta\":{\"mcps\":{}}}}");
        var (store, corrupt) = MasterStoreIO.Load(Url);
        Assert.Null(corrupt);
        Assert.Equal("Alpha", store.ActiveCollection);
    }

    [Fact]
    public void V1FormatFileIsTreatedAsCorruptAndRebuilt()
    {
        File.WriteAllText(Url, "{\"version\":1,\"mcps\":{\"scoutbook\":{\"enabled\":true,\"config\":{\"command\":\"npx\",\"args\":[\"-y\",\"mcp-remote\",\"https://example.com/mcp\"]},\"lastEditView\":\"form\"}}}");
        var (store, corrupt) = MasterStoreIO.Load(Url);
        Assert.Equal(MasterStore.Empty(), store);
        Assert.NotNull(corrupt);
        Assert.StartsWith("mcps.corrupt.", Path.GetFileName(corrupt), StringComparison.Ordinal);
        Assert.False(File.Exists(Url));
    }

    [Theory]
    [InlineData("{\"version\":2,\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{\"s\":{\"enabled\":true,\"config\":{}}}}}}")]                    // lastEditView missing
    [InlineData("{\"version\":2,\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{\"s\":{\"enabled\":\"yes\",\"config\":{},\"lastEditView\":\"form\"}}}}}")]  // enabled not bool
    [InlineData("{\"version\":2,\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{\"s\":{\"enabled\":true,\"config\":{},\"lastEditView\":\"grid\"}}}}}")]   // unknown view
    [InlineData("{\"version\":\"2\",\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{}}}}")]                                                  // version not a number
    [InlineData("{\"version\":2,\"profiles\":{\"D\":{\"mcps\":{}}}}")]                                                                             // activeProfile missing
    [InlineData("[]")]
    public void RequiredKeysAreStrictLikeSwiftCodable(string json)
    {
        Assert.Throws<FormatException>(() => MasterStore.FromJson(JsonValue.Parse(json)));
    }

    [Fact]
    public void UnknownKeysAreIgnored()
    {
        var store = MasterStore.FromJson(JsonValue.Parse("{\"version\":2,\"activeProfile\":\"D\",\"future\":1,\"profiles\":{\"D\":{\"mcps\":{},\"note\":\"x\"}}}"));
        Assert.Equal("D", store.ActiveCollection);
    }

    // mcps accessor scoping

    [Fact]
    public void McpsAccessorReadsAndWritesOnlyActiveCollection()
    {
        var store = TwoCollections();
        store.Mcps["c"] = Entry("https://c.example/mcp");
        Assert.Equal(["a", "c"], Keys(store.Collections["Work"]));
        Assert.Equal(["b"], Keys(store.Collections["Personal"]));
    }

    // Collection management

    [Fact]
    public void AddCollectionCopyingCurrent()
    {
        var store = new MasterStore(new Dictionary<string, McpEntry> { ["a"] = Entry("https://a.example/mcp") });
        Assert.Null(store.AddCollection("Copy", copyingCurrent: true));
        Assert.Equal("Copy", store.ActiveCollection);
        Assert.Equal(["a"], Keys(store.Collections["Copy"]));
        Assert.Equal(["a"], Keys(store.Collections["Default"]));
    }

    [Fact]
    public void AddCollectionEmptyStartsBlank()
    {
        var store = new MasterStore(new Dictionary<string, McpEntry> { ["a"] = Entry("https://a.example/mcp") });
        Assert.Null(store.AddCollection("Fresh", copyingCurrent: false));
        Assert.Empty(store.Collections["Fresh"].Mcps);
    }

    [Fact]
    public void AddCollectionRejectsEmptyName()
    {
        Assert.Equal("Name must not be empty.", MasterStore.Empty().AddCollection("   ", false));
    }

    [Fact]
    public void AddCollectionRejectsDuplicateName()
    {
        Assert.Equal("A collection named “Default” already exists.", MasterStore.Empty().AddCollection("Default", false));
    }

    [Fact]
    public void RenameActiveCollection()
    {
        var store = MasterStore.Empty();
        Assert.Null(store.RenameCollection(store.ActiveCollection, "Main"));
        Assert.Equal("Main", store.ActiveCollection);
        Assert.Equal(["Main"], store.Collections.Keys.ToArray());
    }

    [Fact]
    public void RenameActiveCollectionRejectsCollision()
    {
        var store = new MasterStore(2, "Work", [new("Work", new Collection()), new("Personal", new Collection())]);
        Assert.Equal("A collection named “Personal” already exists.", store.RenameCollection(store.ActiveCollection, "Personal"));
        Assert.Equal("Work", store.ActiveCollection);
    }

    [Fact]
    public void RenameActiveCollectionRejectsEmptyName()
    {
        var store = MasterStore.Empty();
        Assert.NotNull(store.RenameCollection(store.ActiveCollection, "  "));
    }

    [Fact]
    public void DeleteActiveCollectionSwitchesToFirstRemaining()
    {
        var store = new MasterStore(2, "Work", [new("Work", new Collection()), new("Alpha", new Collection()), new("Zeta", new Collection())]);
        Assert.Null(store.DeleteCollection(store.ActiveCollection));
        Assert.Equal("Alpha", store.ActiveCollection);
        Assert.False(store.Collections.ContainsKey("Work"));
    }

    [Fact]
    public void DeleteActiveCollectionRejectsLastCollection()
    {
        var store = MasterStore.Empty();
        Assert.Equal("Can’t delete the last collection.", store.DeleteCollection(store.ActiveCollection));
        Assert.Single(store.Collections);
    }

    [Fact]
    public void SwitchCollection()
    {
        var store = new MasterStore(2, "Work", [new("Work", new Collection()), new("Personal", new Collection())]);
        Assert.Null(store.SwitchCollection("Personal"));
        Assert.Equal("Personal", store.ActiveCollection);
    }

    [Fact]
    public void SwitchCollectionRejectsUnknownName()
    {
        var store = MasterStore.Empty();
        Assert.Equal("No collection named “Nope”.", store.SwitchCollection("Nope"));
        Assert.Equal("Default", store.ActiveCollection);
    }

    [Fact]
    public void ErrorMessagesUseTypographicPunctuationLikeTheMacApp()
    {
        var store = MasterStore.Empty();
        var duplicate = store.AddCollection("Default", false)!;
        Assert.Equal('“', duplicate[duplicate.IndexOf("Default", StringComparison.Ordinal) - 1]);
        Assert.Equal('”', duplicate[duplicate.IndexOf("Default", StringComparison.Ordinal) + "Default".Length]);
        Assert.Contains('’', store.DeleteCollection(store.ActiveCollection)!);
    }

    [Fact]
    public void CloneIsDeepAndEqual()
    {
        var store = TwoCollections();
        var clone = store.Clone();
        Assert.Equal(store, clone);
        clone.Mcps["z"] = Entry("https://z.example/mcp");
        Assert.NotEqual(store, clone);
        Assert.False(store.Mcps.ContainsKey("z"));
    }

    [Fact]
    public void VersionBeyondInt32IsAcceptedLikeSwiftInt64()
    {
        var store = MasterStore.FromJson(JsonValue.Parse("{\"version\":99999999999,\"activeProfile\":\"D\",\"profiles\":{\"D\":{\"mcps\":{}}}}"));
        Assert.Equal(99999999999L, store.Version);
    }
}
